app/scripts/modules/core/cluster/cluster.service.ts

```diff
@@ -0,0 +1,236 @@
+    const groupedByAccount = groupBy(serverGroups, 'account');
+    forOwn(groupedByAccount, (accountServerGroups, account) => {
+      const groupedByCategory = groupBy(accountServerGroups, 'category');
+      forOwn(groupedByCategory, (categoryServerGroups, category) => {
+        const groupedByCluster = groupBy(categoryServerGroups, 'cluster');
+        forOwn(groupedByCluster, (clusterServerGroups, clusterName) => {
+          const cluster: ICluster = {
+            account: account,
+            category: category,
+            name: clusterName,
+            serverGroups: clusterServerGroups,
+            cloudProvider: clusterServerGroups[0].cloudProvider
+          };
+          this.addHealthCountsToCluster(cluster);
+          clusters.push(cluster);
+        });
+      });
+    });
+    this.addProvidersAndServerGroupsToInstances(serverGroups);
+    return clusters;
+  }
+
+  public addExecutionsToServerGroups(application: Application): void {
+    let executions = get(application, 'runningExecutions.data', []);
+
+    if (!application.serverGroups.data) {
+      return; // still run if there are no running tasks, since they may have all finished and we need to clear them.
+    }
+
+    application.serverGroups.data.forEach((serverGroup: ServerGroup) => {
+      serverGroup.runningExecutions = [];
+      executions.forEach((execution: IExecution) => {
+        this.findStagesWithServerGroupInfo(execution.stages).forEach((stage: IExecutionStage) => {
+          const stageServerGroup = stage ? this.extractServerGroupNameFromContext(stage.context) : '';
+          const stageAccount = stage && stage.context ? stage.context.account || stage.context.credentials : '';
+          const stageRegion = stage ? this.extractRegionFromContext(stage.context) : '';
+
+          if (stageServerGroup.includes(serverGroup.name) &&
+            stageAccount === serverGroup.account &&
+            stageRegion === serverGroup.region) {
+              serverGroup.runningExecutions.push(execution);
+          }
+        });
+      });
+    });
+  }
+
+  public addTasksToServerGroups(application: Application): void {
+    let runningTasks = get(application, 'runningTasks.data', []);
+    if (!application.serverGroups.data) {
+      return; // still run if there are no running tasks, since they may have all finished and we need to clear them.
+    }
+    application.serverGroups.data.forEach((serverGroup: ServerGroup) => {
+      if (!serverGroup.runningTasks) {
+        serverGroup.runningTasks = [];
+      }
+      serverGroup.runningTasks.length = 0;
+      runningTasks.forEach(function(task) {
+        if (taskMatches(task, serverGroup)) {
+          serverGroup.runningTasks.push(task);
+        }
+      });
+    });
+  }
+
+  private getClusters(application: string): IPromise<IClusterSummary[]> {
+    return this.API.one('applications').one(application).one('clusters').get().then((clustersMap: {[account: string]: string[]}) => {
```

Comment:
Good catch - I need to work through this. Right now, if this call fails, the page won't load at all, even to indicate that there was an error. I'll get this fixed and updated later this evening.